crates/chia-consensus/src/multiprocess_validation.rs

```diff
@@ -0,0 +1,243 @@
+use crate::consensus_constants::ConsensusConstants;
+use std::sync::{Arc, Mutex};
+use std::thread;
+use crate::gen::condition_tools::pkm_pairs;
+use crate::gen::flags::{
+    AGG_SIG_ARGS, ALLOW_BACKREFS, ENABLE_MESSAGE_CONDITIONS, ENABLE_SOFTFORK_CONDITION,
+    NO_RELATIVE_CONDITIONS_ON_EPHEMERAL,
+};
+use crate::gen::owned_conditions::OwnedSpendBundleConditions;
+use crate::gen::solution_generator::solution_generator;
+use crate::gen::validation_error::ErrorCode;
+use crate::generator_types::BlockGenerator;
+use crate::npc_result::get_name_puzzle_conditions;
+use chia_bls::aggregate_verify;
+use chia_bls::BlsCache;
+use chia_bls::PublicKey;
+use chia_protocol::Coin;
+use chia_protocol::Program;
+use chia_protocol::SpendBundle;
+use chia_traits::Streamable;
+use clvmr::{ENABLE_BLS_OPS_OUTSIDE_GUARD, ENABLE_FIXED_DIV};
+use std::time::{Duration, Instant};
+
+// currently in multiprocess_validation.py
+// called via blockchain.py from full_node.py when a full node wants to add a block or batch of blocks
+fn pre_validate_blocks_multiprocessing() {}
+
+// currently in multiprocess_validation.py
+// called in threads from pre_validate_blocks_multiprocessing
+fn batch_pre_validate_blocks() {}
+
+// currently in mempool_manager.py
+// called in full_node.py when adding a transaction
+pub fn pre_validate_spendbundle(
+    new_spend: SpendBundle,
+    max_cost: u64,
+    constants: ConsensusConstants,
+    peak_height: u32,
+    syncing: bool,
+    cache: Arc<Mutex<BlsCache>>,
```

Comment:
If you need to update it, you can't pass it by immutable reference without interior mutability.

I would say something like `&mut MutexGuard` or `&Mutex` could work though depending on the needs. That way you don't need to pass it an `Arc`.

I would personally recommend using `parking_lot` (sync) or `tokio` (async) for `Mutex` since it can't be poisoned, but don't have strong feelings on it.